src/metabase/sync/sync_metadata/sync_table_privileges.clj

```diff
@@ -0,0 +1,36 @@
+(ns metabase.sync.sync-metadata.sync-table-privileges
+  (:require
+   [clojure.java.jdbc :as jdbc]
+   [honey.sql :as sql]
+   [metabase.driver :as driver]
+   [metabase.driver.util :as driver.u]
+   [metabase.models.interface :as mi]
+   [metabase.util.malli :as mu]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(mu/defn sync-table-privileges!
+  "Sync the `table_privileges` table with the privileges in the database.
+
+   This is a cache of the data returned from `driver/table-privileges`, but it's stored in the database for performance."
+  [database :- (mi/InstanceOf :model/Database)]
+  (let [driver (driver.u/database->driver database)]
+    ;; The following expression should be replaced by (when (driver/supports? :actions driver) ...)
+    ;; when table_privileges are implemented for all the actions drivers
+    (when (contains? (methods driver/current-user-table-privileges) driver)
+      (let [rows               (driver/current-user-table-privileges driver database)
+            schema+table->id   (t2/select-fn->pk (fn [t] {:schema (:schema t), :table (:name t)}) 'Table :db_id (:id database))
+            rows-with-table-id (keep (fn [row]
+                                       (when-let [table-id (get schema+table->id (select-keys row [:schema :table]))]
+                                         (-> row
+                                             (assoc :table_id table-id)
+                                             (dissoc :schema :table))))
+                                     rows)]
+        (t2/with-transaction [conn]
```

Comment:
At the moment, since it's done asynchronously on sync, and it's only for the current DB user, speed hardly matters and it shouldn't take more than a second anyway.

Regardless, the main performance bottleneck is the time it takes to query the data, not deletion/insertion. When we get to support this with connection impersonation the query could take so long on large instances we wouldn't want to update all the table privileges for a database at once. So we'll have to query the privileges by role and on demand. Then we'll have to do a lot more optimization to get it really fast.

Until that point, there's little reason to optimize this.